### PR TITLE
add Moslora

### PR DESCRIPTION
**Pull request by TranscenderNing**
### PR types
New features 

### PR changes
add moslora at peft/lora

### Description
add moslora method

### CODE REVIEW

File: llm/tools/merge_lora_params.py [hunk 1]
```diff
@@ -86,16 +86,25 @@ def lora_process(name, lora_config, state_dict, device, lora_state_dict=None):
         return
 
     weight = state_dict.pop(name + ".weight")
+    lora_use_mixer = (lora_state_dict is not None and name + ".lora_AB" in lora_state_dict) or (
```
Comment (lugimzzz) [inline llm/tools/merge_lora_params.py]: 建议lora_use_mixer直接从lora_config读取不要用state dict的key判断

File: llm/tools/merge_lora_params.py [hunk 2]
```diff
@@ -105,9 +114,16 @@ def lora_process(name, lora_config, state_dict, device, lora_state_dict=None):
         weight = weight.astype("float32")
         lora_A = lora_A.astype("float32")
         lora_B = lora_B.astype("float32")
-        out = (weight + lora_A @ lora_B * scaling).astype("bfloat16")
+        if lora_use_mixer:
+            lora_AB = lora_AB.astype("float32")
+            out = (weight + lora_A @ lora_AB @ lora_B * scaling).astype("bfloat16")
+        else:
+            out = (weight + lora_A @ lora_B * scaling).astype("bfloat16")
```
Comment (lugimzzz) [inline llm/tools/merge_lora_params.py]: 这里astype改成lora_config里的dtype，原本写的有问题

File: llm/utils/argument.py [hunk 3]
```diff
@@ -209,6 +209,9 @@ class ModelArgument:
     rslora: bool = field(default=False, metadata={"help": "Whether to use RsLoRA"})
     lora_plus_scale: float = field(default=1.0, metadata={"help": "Lora B scale in LoRA+ technique"})
     pissa: bool = field(default=False, metadata={"help": "Whether to use Pissa: https://arxiv.org/pdf/2404.02948.pdf"})
+    lora_use_mixer: bool = field(
+        default=False, metadata={"help": "Whether to use MosLoRA: https://arxiv.org/pdf/2406.11909"}
```
Comment (lugimzzz) [inline llm/utils/argument.py:219]: 顺带更新一下文档

File: paddlenlp/peft/lora/lora_model.py [hunk 4]
```diff
@@ -443,6 +443,7 @@ def _find_and_replace_module(self, model, module_name, lora_config, enable_lora)
                 pissa=lora_config.pissa,
                 bias_attr=False if module.bias is None else None,
                 use_quick_lora=lora_config.use_quick_lora,
+                lora_use_mixer=lora_config.lora_use_mixer,
```
Comment (lugimzzz) [inline paddlenlp/peft/lora/lora_model.py:449]: 在loramodel的init里if (tensor_parallel_degree >1 or pipeline_parallel_degree > 1) and lora_config.lora_use_mixer: raise NotImplementError（"xxx"）

File: llm/tools/merge_lora_params.py (hunk 1 above)
Comment (TranscenderNing) [reply llm/tools/merge_lora_params.py]: resoved

File: llm/tools/merge_lora_params.py (hunk 2 above)
Comment (TranscenderNing) [reply llm/tools/merge_lora_params.py]: resolved

File: llm/utils/argument.py (hunk 3 above)
Comment (TranscenderNing) [reply llm/utils/argument.py:219]: resolved

File: paddlenlp/peft/lora/lora_model.py (hunk 4 above)
Comment (TranscenderNing) [reply paddlenlp/peft/lora/lora_model.py:449]: resolved

**Review by lugimzzz — APPROVED** [review]
LGTM

**Pull request merged**